text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1042 @@
+
+```rust
+const fn default<T: [const] Default>() -> T {
+    T::default()
+}
+```
+
+`[const]` means "conditionally" in this context, or specifically "const impl required if called in const context".
+It is the opposite of `?` (preexisting for `?Sized` bounds), which also means "conditionally", but from the other direction: `?const`
+(not proposed here, see [this alternatives section](#make-all-const-fn-arguments-const-trait-by-default-and-require-an-opt-out-const-trait) for why it was rejected)
+would mean "no const impl required, even if called in const context".
+
+### Const fn
+
+`const` fn have always been and will stay "always const" functions.
+
+It may appear that a function is suddenly "not a const fn" if it gets passed a type that doesn't satisfy
+the constness of the corresponding trait bound. E.g.
+
+```rust
+struct Foo;
+
+impl Clone for Foo {
+    fn clone(&self) -> Self {
+        Foo
+    }
+}
+
+const fn bar<T: [const] Clone>(t: &T) -> T { t.clone() }
+const BAR: Foo = bar(Foo); // ERROR: `Foo`'s `Clone` impl is not for `const Clone`.
+```
+
+But `bar` is still a `const` fn and you can call it from a const context, it will just fail some trait bounds. This is no different from 
+
+```rust
+const fn dup<T: Copy>(a: T) -> (T, T) {(a, a)}
+const FOO: (String, String) = dup(String::new());
+```
+
+Here `dup` is always const fn, you'll just get a trait bound failure if the type you pass isn't `Copy`.
+
+This may seem like language lawyering, but that's how the impl works and how I believe we should be talking about it.
+
+It's actually important for inference and method resolution in the nonconst world today.
+You first figure out which method you're calling, then you check its bounds.
+Otherwise it would at least seem like we'd have to allow some SFINAE or method overloading style things,
+which we definitely do not support and have historically rejected over and over again.
+
+### Impls with const methods for conditionally const trait methods
+
+`const trait` impls for generic types work similarly to generic `const fn`.
+Any `impl const Trait for Type` is allowed to have `[const]` trait bounds:
+
+```rust
+struct MyStruct<T>(T);
+
+impl<T: [const] Add<Output = T>> const Add for MyStruct<T> {
+    type Output = MyStruct<T>;
+    fn add(self, other: MyStruct<T>) -> MyStruct<T> {
+        MyStruct(self.0 + other.0)
+    }
+}
+
+impl<T> const Add for &MyStruct<T>
+where
+    for<'a> &'a T: [const] Add<Output = T>,
+{
+    type Output = MyStruct<T>;
+    fn add(self, other: &MyStruct<T>) -> MyStruct<T> {
+        MyStruct(&self.0 + &other.0)
+    }
+}
+```
+
+See [this playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=313a38ef5c36b2ddf489f74167c1ac8a) for an example that works on nightly today.
+
+### Derives
+
+Most of the time you don't want to write out your impls by hand, but instead derive them as the implementation is obvious from your data structure.
+
+```rust
+#[const_derive(PartialEq, Eq)]
```

Comment:
This looks like a typo - you talk below (consistently) about `derive_const`, but this is `const_derive`.